wfe/wfe.go

```diff
@@ -783,12 +786,36 @@ func (wfe *WebFrontEndImpl) NewCertificate(ctx context.Context, logEvent *reques
 		return
 	}
 
+	var rawCSR core.RawCertificateRequest
+	err := json.Unmarshal(body, &rawCSR)
+	if err != nil {
+		logEvent.AddError("unable to JSON unmarshal CertificateRequest: %s", err)
+		wfe.sendError(response, logEvent, probs.Malformed("Error unmarshaling certificate request"), err)
+		return
+	}
+	if wfe.CheckMalformedCSR {
+		// Assuming a properly formatted CSR there should be two four byte SEQUENCE
+		// declarations then a two byte integer declaration which defines the version
+		// of the CSR. If those two bytes (at offset 8 and 9) and equal to 2 and 0
+		// then the CSR was generated by a pre-1.0.2 version of OpenSSL with a client
+		// which didn't explicitly set the version causing the integer to be malformed
+		// and encoding/asn1 will refuse to parse it. If this is the case exit early
+		// with a more useful error message.
+		if len(rawCSR.CSR) >= 10 && rawCSR.CSR[8] == 2 && rawCSR.CSR[9] == 0 {
+			logEvent.AddError("Pre-1.0.2 OpenSSL malformed CSR")
+			wfe.sendError(response, logEvent, probs.Malformed("CSR generated using a pre-1.0.2 OpenSSL with a client that doesn't properly specify the CSR version"), nil)
+			return
+		}
+	}
+
 	var certificateRequest core.CertificateRequest
-	if err := json.Unmarshal(body, &certificateRequest); err != nil {
+	certificateRequest.CSR, err = x509.ParseCertificateRequest(rawCSR.CSR)
+	if err != nil {
 		logEvent.AddError("unable to JSON unmarshal CertificateRequest: %s", err)
```

Comment:
This was the error message being used for the `err` produced by `json.Unmarshal`. With the refactor in this PR the `err` will be produced by `x509.ParseCertificateRequest` and does't appear to relate to JSON unmarshaling anymore. Does the log message/`probs.Malformed` here need updating?